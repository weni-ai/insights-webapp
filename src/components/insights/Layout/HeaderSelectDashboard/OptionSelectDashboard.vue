<template>
  <UnnnicDropdownItem
    data-testid="option-select-dashboard"
    class="option-select-dashboard"
    :class="{
      'option-select-dashboard--active':
        currentDashboard.uuid === dashboard.uuid,
    }"
    @click="setCurrentDashboard(dashboard)"
  >
    {{ dashboard.name }}
    <UnnnicIcon
      data-testid="star-icon"
      class="option-select-dashboard__star-icon"
      :class="{
        'option-select-dashboard__star-icon--selected': isDefaultDashboard,
      }"
      icon="star_rate"
      scheme="neutral-clean"
      clickable
      :filled="isDefaultDashboard || starHovered"
      @mouseenter="setStarHovered(true)"
      @mouseleave="setStarHovered(false)"
      @click.stop="handleSetDefaultDashboard"
    />
  </UnnnicDropdownItem>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import Unnnic from '@weni/unnnic-system';

export default {
  name: 'OptionSelectDashboard',

  props: {
    dashboard: {
      type: Object,
      default: () => {},
    },
  },

  data() {
    return {
      dashboardHovered: '',
      starHovered: false,
    };
  },

  computed: {
    ...mapState({
      currentDashboard: (state) => state.dashboards.currentDashboard,
    }),
    ...mapGetters({
      dashboardDefault: 'dashboards/dashboardDefault',
    }),

    isDefaultDashboard() {
      return this.dashboardDefault?.uuid === this.dashboard.uuid;
    },
  },

  methods: {
    ...mapActions({
      setCurrentDashboard: 'dashboards/setCurrentDashboard',
      setDefaultDashboard: 'dashboards/setDefaultDashboard',
    }),

    setStarHovered(boolean) {
      this.starHovered = boolean;
    },

    async handleSetDefaultDashboard() {
      const { dashboard } = this;
      if (dashboard.uuid === this.dashboardDefault.uuid) return;

      try {
        await this.setDefaultDashboard(dashboard.uuid);
        this.callSetDashboardAlert('success');
      } catch (error) {
        console.error(error);
        this.callSetDashboardAlert('error');
      }
    },

    callSetDashboardAlert(type) {
      const { dashboard } = this;

      if (!['success', 'error'].includes(type)) {
        throw new Error(
          'Error calling the alert when setting the default dashboard. This type does not exist.',
        );
      }

      Unnnic.unnnicCallAlert({
        props: {
          text: this.$t(`insights_header.set_default_dashboard_${type}`, {
            dashboard: dashboard.name,
          }),
          type,
        },
        seconds: 5,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
.option-select-dashboard {
  &.unnnic-dropdown-item::before {
    content: none;
  }

  border-radius: $unnnic-border-radius-sm;

  padding: $unnnic-spacing-xs;

  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: $unnnic-spacing-nano;

  color: $unnnic-color-neutral-darkest;
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-gt;

  &--active {
    background-color: $unnnic-color-neutral-lightest;
    font-weight: $unnnic-font-weight-bold;
  }

  .option-select-dashboard__star-icon:not(
      .option-select-dashboard__star-icon--selected
    ):hover {
    color: $unnnic-color-weni-500;
  }

  .option-select-dashboard__star-icon--selected {
    color: $unnnic-color-weni-600;
  }
}
</style>
