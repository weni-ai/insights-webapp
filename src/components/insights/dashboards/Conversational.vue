<template>
  <section class="dashboard-conversational">
    <DashboardHeader class="dashboard-conversational__header" />

    <MostTalkedAboutTopicsWidget
      class="dashboard-conversational__most-talked-about-topics"
    />

    <ConversationalDynamicWidget
      v-for="(widget, index) in orderedDynamicWidgets"
      :key="index"
      :type="widget.type"
      :uuid="widget.uuid"
      class="dashboard-conversational__dynamic-widget"
      :class="{
        'dashboard-conversational__dynamic-widget--only-add': isOnlyAddWidget(
          widget.type,
        ),
      }"
    />

    <CustomizableDrawer />
  </section>
</template>

<script setup lang="ts">
import { computed, onUnmounted, ref, watch } from 'vue';
import { storeToRefs } from 'pinia';
import DashboardHeader from '@/components/insights/conversations/DashboardHeader.vue';
import MostTalkedAboutTopicsWidget from '@/components/insights/conversations/MostTalkedAboutTopicsWidget/index.vue';
import ConversationalDynamicWidget from '@/components/insights/conversations/ConversationalDynamicWidget.vue';
import { useConversationalWidgets } from '@/store/modules/conversational/widgets';
import { useWidgets } from '@/store/modules/widgets';
import CustomizableDrawer from '@/components/insights/conversations/CustomizableWidget/CustomizableDrawer.vue';
import { useCustomWidgets } from '@/store/modules/conversational/customWidgets';

type ConversationalWidgetType = 'csat' | 'nps' | 'add' | 'custom';

const customWidgets = useCustomWidgets();
const { getCustomWidgets } = storeToRefs(customWidgets);

const conversationalWidgets = useConversationalWidgets();
const { isCsatConfigured, isNpsConfigured, getDynamicWidgets } = storeToRefs(
  conversationalWidgets,
);
const widgets = useWidgets();
const { isLoadingCurrentDashboardWidgets, currentDashboardWidgets } =
  storeToRefs(widgets);

const dynamicWidgets = ref<{ type: ConversationalWidgetType; uuid: string }[]>(
  [],
);

const orderedDynamicWidgets = computed(() => {
  if (isLoadingCurrentDashboardWidgets.value) {
    return [];
  }

  return dynamicWidgets.value;
});

const isOnlyAddWidget = (widget: ConversationalWidgetType) => {
  const isOddNumberOfWidgets = dynamicWidgets.value.length % 2 === 1;
  const isLastWidgetAdd = widget === 'add';

  return isOddNumberOfWidgets && isLastWidgetAdd;
};

watch(currentDashboardWidgets, () => {
  if (isCsatConfigured.value) {
    dynamicWidgets.value.push({ type: 'csat', uuid: '' });
  }

  if (isNpsConfigured.value) {
    dynamicWidgets.value.push({ type: 'nps', uuid: '' });
  }

  if (getCustomWidgets.value.length > 0) {
    getCustomWidgets.value.forEach((widget) => {
      dynamicWidgets.value.push({ type: 'custom', uuid: widget.uuid });
    });
  }

  dynamicWidgets.value.push({ type: 'add', uuid: '' });
});

onUnmounted(() => {
  dynamicWidgets.value = [];
});
</script>

<style lang="scss" scoped>
$layout-gap: $unnnic-spacing-sm;

.dashboard-conversational {
  display: flex;
  gap: $layout-gap;
  flex-wrap: wrap;

  &__header {
    width: 100%;
  }

  &__most-talked-about-topics {
    width: 100%;
  }

  &__dynamic-widget {
    min-width: calc(50% - $layout-gap / 2);
    width: fit-content;

    &--only-add {
      min-width: 100%;
    }
  }
}
</style>
