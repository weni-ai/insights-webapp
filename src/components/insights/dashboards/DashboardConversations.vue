<template>
  <section class="dashboard-conversations">
    <TotalConversationsWidget />
    <section class="dashboard-conversations__widget-issues-container">
      <TopicsWidget class="dashboard-conversations__topics-widget" />
      <QueueServiceWidget
        class="dashboard-conversations__queue-service-widget"
      />
    </section>
  </section>
  <section v-if="isLoading">Loading...</section>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import TotalConversationsWidget from '@/components/insights/widgets/Conversations/TotalConversationsWidget.vue';
import QueueServiceWidget from '@/components/insights/widgets/Conversations/QueueServiceWidget.vue';
import TopicsWidget from '@/components/insights/widgets/Conversations/TopicsWidget.vue';

const isLoading = ref(false);
</script>

<style lang="scss" scoped>
.dashboard-conversations {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__widget-issues-container {
    display: flex;
    gap: $unnnic-spacing-md;
    width: 100%;
    height: 100%;
  }

  &__topics-widget {
    flex: 6;
  }

  &__queue-service-widget {
    flex: 4;
  }
}
</style>
