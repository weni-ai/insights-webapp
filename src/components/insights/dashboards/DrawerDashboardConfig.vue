<template>
  <UnnnicDrawer
    ref="unnnicDrawer"
    :style="{ display: showProgressBar ? 'none' : 'flex' }"
    :modelValue="modelValue"
    :title="dashboard ? $t('edit_dashboard.title') : $t('new_dashboard.title')"
    :primaryButtonText="$t('save')"
    :disabledPrimaryButton="!isValidConfig"
    :loadingPrimaryButton="loadingRequest"
    :secondaryButtonText="$t('cancel')"
    :disabledSecondaryButton="loadingRequest"
    wide
    @primary-button-click="dashboard ? updateDashboard() : createDashboard()"
    @secondary-button-click="$refs.unnnicDrawer.close()"
    @close="close"
  >
    <template #content>
      <form
        class="config-form"
        @submit.prevent
      >
        <section class="config-form__input">
          <UnnnicLabel :label="$t('dashboard_name')" />
          <UnnnicInput
            v-model="dashboardForm.name"
            :placeholder="$t('new_dashboard.dashboard_name_placeholder')"
          />
        </section>
        <section
          v-if="!dashboard"
          class="config-form__input"
        >
          <UnnnicLabel :label="$t('funnel')" />
          <UnnnicSelectSmart
            v-model="dashboardForm.qtdFunnel"
            :options="funnelOptions"
            :placeholder="$t('select')"
          />
          <p class="config-form__input-hint">
            {{ $t('new_dashboard.funnel_hint') }}
          </p>
        </section>
        <section class="config-form__input">
          <UnnnicLabel :label="$t('currency')" />
          <UnnnicSelectSmart
            v-model="dashboardForm.currency"
            :options="currencyOptions"
            :placeholder="$t('select')"
          />
        </section>
        <UnnnicButton
          v-if="dashboard?.is_deletable"
          class="config-form__delete-dashboard-button"
          type="tertiary"
          :text="$t('edit_dashboard.delete_dashboard')"
          @click="showDeleteDashboardModal = true"
        />
      </form>
    </template>
  </UnnnicDrawer>
  <ProgressBar
    v-if="showProgressBar"
    :title="$t('new_dashboard.creating_new_dashboard')"
    @progress-complete="handleCreateProgressComplete"
  />
  <ModalDeleteDashboard
    v-if="showDeleteDashboardModal"
    v-model="showDeleteDashboardModal"
    :dashboard="dashboard"
    @close="$event.cascade ? close() : (showDeleteDashboardModal = false)"
  />
</template>

<script>
import { mapMutations, mapState } from 'vuex';
import unnnic from '@weni/unnnic-system';

import ProgressBar from '@/components/ProgressBar.vue';
import ModalDeleteDashboard from './ModalDeleteDashboard.vue';

import { Dashboards } from '@/services/api';
import { Dashboard } from '@/models';
export default {
  name: 'DrawerDashboardConfig',
  components: { ProgressBar, ModalDeleteDashboard },
  props: {
    modelValue: {
      type: Boolean,
      required: true,
    },
    dashboard: {
      type: [Object, undefined],
      default: undefined,
    },
  },
  emits: ['close', 'update:modelValue'],
  data() {
    return {
      dashboardForm: {
        name: '',
        qtdFunnel: [{ label: '1', value: '1' }],
        currency: [],
      },
      funnelOptions: [
        { label: this.$t('select'), value: '' },
        { label: this.$t('none'), value: '0' },
        { label: '1', value: '1' },
        { label: '2', value: '2' },
        { label: '3', value: '3' },
      ],
      currencyOptions: [
        { label: this.$t('currency_options.BRL'), value: 'BRL' },
        { label: this.$t('currency_options.USD'), value: 'USD' },
        { label: this.$t('currency_options.EUR'), value: 'EUR' },
        { label: this.$t('currency_options.ARS'), value: 'ARS' },
      ],
      loadingRequest: false,
      createdDashboard: {},
      showProgressBar: false,

      showDeleteDashboardModal: false,
    };
  },
  computed: {
    ...mapState({ dashboards: (state) => state.dashboards.dashboards }),

    isValidConfig() {
      const commonValidations = !!(
        this.dashboardForm.name.trim() &&
        this.dashboardForm.currency.length &&
        this.dashboardForm.currency[0].value
      );

      if (!this.dashboard) {
        return !!(
          commonValidations &&
          this.dashboardForm.qtdFunnel.length &&
          this.dashboardForm.qtdFunnel[0].value !== ''
        );
      }

      return commonValidations;
    },
  },
  mounted() {
    if (this.dashboard) {
      this.handleDashboardFields();
    }
  },
  methods: {
    ...mapMutations({
      setDashboards: 'dashboards/SET_DASHBOARDS',
      setCurrentDashboard: 'dashboards/SET_CURRENT_DASHBOARD',
    }),
    handleDashboardFields() {
      const currencyOption = this.currencyOptions.find(
        (currency) => currency.value === this.dashboard.config?.currency_type,
      );
      this.dashboardForm.currency = currencyOption ? [currencyOption] : [];
      this.dashboardForm.name = this.dashboard.name;
    },
    close() {
      this.$emit('close');
    },
    handleCreateProgressComplete() {
      this.loadingRequest = false;
      this.dashboards.push(this.createdDashboard);
      this.setCurrentDashboard(this.createdDashboard);
      this.$router.push({
        name: 'dashboard',
        params: {
          dashboardUuid: this.createdDashboard.uuid,
        },
      });
      unnnic.unnnicCallAlert({
        props: {
          text: this.$t('new_dashboard.alert.success'),
          type: 'success',
        },
        seconds: 5,
      });
      this.close();
    },
    createDashboard() {
      this.loadingRequest = true;

      Dashboards.createFlowsDashboard({
        dashboardName: this.dashboardForm.name,
        funnelAmount: Number(this.dashboardForm.qtdFunnel[0].value),
        currencyType: this.dashboardForm.currency[0].value,
      })
        .then((response) => {
          const { dashboard } = response;
          this.createdDashboard = new Dashboard(
            dashboard.uuid,
            dashboard.name,
            { columns: dashboard.grid[0], rows: dashboard.grid[1] },
            dashboard.is_default,
            dashboard.is_editable,
            dashboard.is_deletable,
            dashboard.config,
          );
          this.showProgressBar = true;
        })
        .catch((error) => {
          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('new_dashboard.alert.error'),
              type: 'error',
            },
            seconds: 5,
          });
          console.log(error);
          this.close();
        });
    },
    updateDashboard() {
      this.loadingRequest = true;

      Dashboards.updateFlowsDashboard({
        dashboardUuid: this.dashboard.uuid,
        dashboardName: this.dashboardForm.name,
        currencyType: this.dashboardForm.currency[0].value,
      })
        .then((response) => {
          let updatedDashboard;
          const dashboards = this.dashboards.map((dash) => {
            if (dash.uuid === this.dashboard.uuid) {
              updatedDashboard = {
                ...dash,
                name: response.name,
                config: {
                  ...dash.config,
                  currency_type: response.config.currency_type,
                },
              };
              return updatedDashboard;
            }
            return dash;
          });

          this.setDashboards(dashboards);
          this.setCurrentDashboard(updatedDashboard);

          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('edit_dashboard.alert.success'),
              type: 'success',
            },
            seconds: 5,
          });
        })
        .catch((error) => {
          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('edit_dashboard.alert.error'),
              type: 'error',
            },
            seconds: 5,
          });
          console.log(error);
        })
        .finally(() => {
          this.loadingRequest = false;
          this.close();
        });
    },
  },
};
</script>

<style lang="scss" scoped>
.unnnic-label__label:first-child {
  margin-top: 0;
}

.config-form {
  display: grid;
  gap: $unnnic-spacing-sm;
  &__input-hint {
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-md;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-line-height-large * 1.25; // 20px
    margin-top: $unnnic-spacing-nano;
  }
}

// Temporary adjustments to allow dropdowns to not be limited to the modal space and may overflow
:deep(.unnnic-select-smart__options.active) {
  position: fixed;
  left: auto;
  right: auto;
}
</style>
