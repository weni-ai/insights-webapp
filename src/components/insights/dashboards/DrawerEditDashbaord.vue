<template>
  <UnnnicDrawer
    ref="unnnicDrawer"
    :modelValue="modelValue"
    :title="$t('new_dashboard.title')"
    :primaryButtonText="$t('save')"
    :disabledPrimaryButton="!isValidConfig"
    :loadingPrimaryButton="loadingRequest"
    :secondaryButtonText="$t('cancel')"
    :disabledSecondaryButton="loadingRequest"
    wide
    @primary-button-click="updateDashboard"
    @secondary-button-click="$refs.unnnicDrawer.close()"
    @close="close"
  >
    <template #content>
      <form @submit.prevent>
        <UnnnicLabel :label="$t('dashboard_name')" />
        <UnnnicInput
          v-model="dashboardConfig.name"
          :placeholder="$t('new_dashboard.dashboard_name_placeholder')"
        />
        <UnnnicLabel :label="$t('currency')" />
        <UnnnicSelectSmart
          v-model="dashboardConfig.currency"
          :options="currencyOptions"
          :placeholder="$t('select')"
        />
      </form>
    </template>
  </UnnnicDrawer>
</template>

<script>
import { Dashboards } from '@/services/api';
import unnnic from '@weni/unnnic-system';
export default {
  name: 'DrawerEditDashboard',
  props: {
    modelValue: {
      type: Boolean,
      required: true,
    },
    dashboard: { type: Object, required: true },
  },
  emits: ['close', 'update:modelValue'],
  data() {
    return {
      dashboardConfig: {
        name: '',
        currency: [],
      },
      currencyOptions: [
        { label: this.$t('currency_options.BRL'), value: 'BRL' },
        { label: this.$t('currency_options.USD'), value: 'USD' },
        { label: this.$t('currency_options.EUR'), value: 'EUR' },
        { label: this.$t('currency_options.ARS'), value: 'ARS' },
      ],
      loadingRequest: false,
    };
  },
  computed: {
    isValidConfig() {
      return (
        this.dashboard.name.trim() &&
        this.dashboard.currency.length &&
        this.dashboard.currency[0].value
      );
    },
  },
  methods: {
    close() {
      this.$emit('close');
    },
    updateDashboard() {
      this.loadingRequest = true;

      Dashboards.createFlowsDashboard({
        dashboardName: this.dashboardConfig.name,
        currencyType: this.dashboardConfig.currency[0].value,
      })
        .then((response) => {
          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('new_dashboard.alert.success'),
              type: 'success',
            },
            seconds: 5,
          });
          // TODO redirect
          this.close();
        })
        .catch((error) => {
          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('new_dashboard.alert.error'),
              type: 'error',
            },
            seconds: 5,
          });
          console.log(error);
        })
        .finally(() => {
          this.loadingRequest = false;
        });
    },
  },
};
</script>

<style lang="scss" scoped>
.unnnic-label__label:first-child {
  margin-top: 0;
}
.input-hint {
  font-family: $unnnic-font-family-secondary;
  font-size: $unnnic-font-size-body-md;
  font-weight: $unnnic-font-weight-regular;
  line-height: $unnnic-line-height-large * 1.25; // 20px
  margin-top: $unnnic-spacing-nano;
}

// Temporary adjustments to allow dropdowns to not be limited to the modal space and may overflow
:deep(.unnnic-select-smart__options.active) {
  position: fixed;
  left: auto;
  right: auto;
}
</style>
