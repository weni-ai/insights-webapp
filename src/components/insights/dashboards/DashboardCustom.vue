<template>
  <section
    class="dashboard"
    :style="dashboardGridStyle"
  >
    <WidgetOnboarding
      v-if="showConfigWidgetOnboarding"
      :showCardTour="showOnboarding.card"
      :showWidgetTour="showOnboarding.empty_widget"
    />
    <section
      v-if="isLoadingCurrentDashboardWidgets"
      class="dashboard__loading"
    >
      <IconLoading />
    </section>
    <template
      v-for="widget of currentDashboardWidgets"
      v-else
      :key="widget.uuid"
    >
      <DynamicWidget
        :style="getWidgetStyle(widget.grid_position)"
        :widget="widget"
        :data-onboarding-id="getWidgetOnboardingId(widget)"
        @open-config="handlerWidgetOpenConfig(widget)"
        @click-data="openFlowResultContactList"
      />
    </template>

    <DrawerConfigGallery
      v-if="!!currentWidgetEditing"
      :modelValue="!!currentWidgetEditing"
      @close="updateCurrentWidgetEditing(null)"
    />
    <FlowResultContactListModal
      v-if="showFlowResultsContactListModal"
      :flowResultLabel="flowResultsContactListParams?.label"
      :flow="flowResultsContactListParams?.flow"
      @close="closeFlowResultContactList()"
    />
  </section>
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapActions, mapState } from 'pinia';

import { useDashboards } from '@/store/modules/dashboards';
import { useWidgets } from '@/store/modules/widgets';
import { useOnboarding } from '@/store/modules/onboarding';

import DrawerConfigGallery from '@/components/insights/drawers/DrawerConfigGallery/index.vue';
import IconLoading from '@/components/IconLoading.vue';
import WidgetOnboarding from '@/components/insights/onboardings/WidgetOnboarding.vue';
import FlowResultContactListModal from '@/components/FlowResultContactListModal.vue';

export default {
  name: 'DashboardCustom',

  components: {
    DynamicWidget: defineAsyncComponent(
      () => import('@/components/insights/widgets/DynamicWidget.vue'),
    ),
    DrawerConfigGallery,
    IconLoading,
    WidgetOnboarding,
    FlowResultContactListModal,
  },

  data() {
    return {
      showDrawerConfigWidget: false,
      widgetConfigurating: null,
      showOnboarding: {
        card: false,
        empty_widget: false,
      },
      showFlowResultsContactListModal: false,
      flowResultsContactListParams: null,
    };
  },

  computed: {
    ...mapState(useDashboards, ['currentDashboard']),
    ...mapState(useWidgets, [
      'currentDashboardWidgets',
      'currentWidgetEditing',
      'isLoadingCurrentDashboardWidgets',
    ]),
    ...mapState(useOnboarding, ['showConfigWidgetOnboarding']),

    isCustomDashboard() {
      return this.currentDashboard.is_deletable;
    },

    dashboardGridStyle() {
      const { grid } = this.currentDashboard || {};
      if (grid) {
        return {
          gridTemplateColumns: `repeat(${grid.columns}, 1fr)`,
          gridTemplateRows: `repeat(${grid.rows}, 1fr)`,
        };
      }
      return {};
    },
  },

  watch: {
    currentDashboardWidgets: {
      immediate: true,
      handler(newCurrentDashboardWidgets) {
        if (
          !!newCurrentDashboardWidgets &&
          this.isCustomDashboard &&
          !this.isLoadingCurrentDashboardWidgets
        ) {
          this.handlerWidgetsOnboarding();
        }
      },
    },
  },

  methods: {
    ...mapActions(useWidgets, ['updateCurrentWidgetEditing']),
    ...mapActions(useOnboarding, [
      'callTourNextStep',
      'setShowConfigWidgetsOnboarding',
    ]),

    openFlowResultContactList(data) {
      this.flowResultsContactListParams = data;
      this.showFlowResultsContactListModal = true;
    },

    closeFlowResultContactList() {
      this.flowResultsContactListParams = {};
      this.showFlowResultsContactListModal = false;
    },

    handleWidgetFilledData() {
      const hasCard = this.currentDashboardWidgets.filter(
        (e) => e.type === 'card',
      );

      const existFunnel = this.currentDashboardWidgets.some(
        (e) => e.type === 'graph_funnel',
      );
      const existVtex = this.currentDashboardWidgets.some(
        (e) => e.type === 'vtex_order',
      );

      this.showOnboarding = {
        card:
          hasCard.length > 0
            ? !!hasCard.every((widget) => widget.name === '')
            : false,
        empty_widget:
          !!this.currentDashboardWidgets.some(
            (widget) => widget.type === 'empty_column',
          ) &&
          !existFunnel &&
          !existVtex,
      };
    },

    handlerWidgetsOnboarding() {
      const hasWidgetsOnboardingComplete =
        localStorage.getItem('hasWidgetsOnboardingComplete') === 'true';

      if (!hasWidgetsOnboardingComplete) {
        this.handleWidgetFilledData();

        if (!this.showOnboarding.card && !this.showOnboarding.empty_widget) {
          localStorage.setItem('hasWidgetsOnboardingComplete', 'true');
        }

        if (this.showOnboarding.card || this.showOnboarding.empty_widget) {
          this.setShowConfigWidgetsOnboarding(true);
        }
      }
    },

    handlerWidgetOpenConfig(widget) {
      const isNewWidget = this.currentWidgetEditing?.uuid !== widget.uuid;
      if (isNewWidget) {
        this.updateCurrentWidgetEditing(widget).then(() => {
          this.callTourNextStep('widgets-onboarding-tour');
        });
      }
    },

    getWidgetStyle(gridPosition) {
      return {
        gridColumn: `${gridPosition.column_start} / ${gridPosition.column_end + 1}`,
        gridRow: `${gridPosition.row_start} / ${gridPosition.row_end + 1}`,
      };
    },

    getWidgetOnboardingId(widget) {
      return widget.type === 'card'
        ? 'widget-card-metric'
        : 'widget-graph-empty';
    },
  },
};
</script>

<style lang="scss" scoped>
.dashboard {
  overflow: hidden;

  height: 100%;

  display: grid;
  gap: $unnnic-spacing-sm;
  &__loading {
    width: 100vw;
    height: 85vh;
    display: flex;
    justify-content: center;
    align-items: center;
  }
}
</style>
