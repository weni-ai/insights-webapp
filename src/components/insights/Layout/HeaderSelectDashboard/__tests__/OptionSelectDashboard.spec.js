import { beforeEach, describe, expect, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

import OptionSelectDashboard from '@/components/insights/Layout/HeaderSelectDashboard/OptionSelectDashboard.vue';
import Unnnic from '@weni/unnnic-system';
import { useDashboards } from '@/store/modules/dashboards';

const dashboard1 = { name: 'Dashboard 1', uuid: '1' };
const dashboard2 = { name: 'Dashboard 2', uuid: '2' };

describe('OptionSelectDashboard', () => {
  let wrapper;
  const store = createTestingPinia({
    initialState: {
      config: { enableCreateCustomDashboards: false },
      dashboards: {
        setDefaultDashboard: vi.fn(() => Promise.resolve()),
        dashboardDefault: () => dashboard1,
        dashboards: [dashboard1, dashboard2],
        currentDashboard: dashboard1,
      },
    },
  });

  const createWrapper = (props) => {
    return mount(OptionSelectDashboard, {
      props,
      global: {
        plugins: [store],
        stubs: {
          UnnnicDropdownItem: Unnnic.unnnicDropdownItem,
          UnnnicIcon: true,
        },
        mocks: {
          $t: (msg) => msg,
        },
      },
    });
  };

  let starIcon;

  beforeEach(() => {
    wrapper = createWrapper({ dashboard: dashboard1 });
    starIcon = wrapper.findComponent('[data-testid=star-icon]');
  });

  describe('Rendering and Interactions', () => {
    it('Should display the name of the dashboard prop', () => {
      const optionSelectDashboard = wrapper.findComponent(
        '[data-testid=option-select-dashboard]',
      );

      expect(optionSelectDashboard.exists()).toBe(true);
      expect(optionSelectDashboard.text()).toContain('Dashboard 1');
    });

    it('Should set the current dashboard when an option is clicked', async () => {
      wrapper = createWrapper({
        dashboard: dashboard2,
      });

      const dashboardsStore = useDashboards();

      dashboardsStore.setCurrentDashboard = (dash) =>
        (dashboardsStore.currentDashboard = dash);

      const optionSelectDashboard = wrapper.findComponent(
        '[data-testid=option-select-dashboard]',
      );
      await optionSelectDashboard.trigger('click');

      expect(dashboardsStore.currentDashboard.uuid).toBe('2');
    });

    it('Should have a star_rate icon', () => {
      expect(starIcon.exists()).toBe(true);
      expect(starIcon.props('icon')).toBe('star_rate');
    });
  });

  describe('Default Dashboard Functionality', () => {
    it('Should set default dashboard when star icon and dashboard is different from default', async () => {
      const spySetDefaultDashboard = vi.spyOn(
        wrapper.vm,
        'setDefaultDashboard',
      );

      await wrapper.setProps({ dashboard: dashboard2 });

      await starIcon.trigger('click');

      expect(spySetDefaultDashboard).toHaveBeenCalled();
    });

    it('Should show a success alert if the default dashboard is set', async () => {
      await wrapper.setProps({ dashboard: dashboard2 });
      const spyUnnnicAlert = vi
        .spyOn(Unnnic, 'unnnicCallAlert')
        .mockImplementation(() => {});
      const createAlertMsg = (type) =>
        wrapper.vm.$t(`insights_header.set_default_dashboard_${type}`, {
          dashboard: dashboard2.name,
        });

      wrapper.vm.setDefaultDashboard = vi.fn(() => Promise.reject());
      await starIcon.trigger('click');
      expect(spyUnnnicAlert).toHaveBeenCalledWith({
        props: {
          text: createAlertMsg('error'),
          type: 'error',
        },
        seconds: 5,
      });

      wrapper.vm.setDefaultDashboard = vi.fn(() => Promise.resolve());
      await starIcon.trigger('click');
      expect(spyUnnnicAlert).toHaveBeenCalledWith({
        props: {
          text: createAlertMsg('success'),
          type: 'success',
        },
        seconds: 5,
      });
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
