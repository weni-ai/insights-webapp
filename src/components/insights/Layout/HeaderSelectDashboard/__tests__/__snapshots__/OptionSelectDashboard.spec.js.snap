// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OptionSelectDashboard > Should match the snapshot 1`] = `
<a
  class="unnnic-dropdown-item unnnic--clickable option-select-dashboard"
  data-testid="option-select-dashboard"
  data-v-000aa0c0=""
>
  
  Dashboard 1 
  <span
    class="material-symbols-rounded unnnic-icon-scheme--neutral-clean unnnic-icon-size--md unnnic--clickable option-select-dashboard__star-icon"
    data-testid="star-icon"
    data-v-000aa0c0=""
    data-v-9846e6db=""
    translate="no"
  >
    star_rate
  </span>
  
</a>
`;
